src/h5web/providers/context.ts

```diff
@@ -10,16 +16,45 @@ export type GetValueParams = {
   selection?: string;
 };
 
-export abstract class ProviderAPI {
-  abstract filepath: string;
-  abstract getEntity(path: string): Promise<Entity>;
-  abstract getValue(params: GetValueParams): Promise<unknown>;
+export abstract class ProviderApi {
+  public readonly filepath: string;
+  protected readonly client: AxiosInstance;
+  protected readonly cancelSources = new Set<CancelTokenSource>();
+
+  public constructor(filepath: string, config?: AxiosRequestConfig) {
+    this.filepath = filepath;
+    this.client = axios.create(config);
+  }
+
+  public cancel(): void {
+    this.cancelSources.forEach((source) => {
+      source.cancel(ProviderError.Cancelled);
+    });
+    this.cancelSources.clear();
+  }
+
+  public async withCancelToken<R>(
+    asyncFn: (cancelToken: CancelToken) => Promise<R>
+  ): Promise<R> {
+    const cancelSource = axios.CancelToken.source();
+    this.cancelSources.add(cancelSource);
+
+    try {
+      return await asyncFn(cancelSource.token);
+    } finally {
+      this.cancelSources.delete(cancelSource);
```

Comment:
We create one cancel source per request, so we need to clean it up if the request succeeds or fails. Reading this again, I wonder if I could have implemented this with a global, unique `cancelSource`...

Reply:
I don't see how: you might have several requests in parallel, right ? Which means several `cancelSource/Token` ?

Reply:
Yeah, but they could all reuse the same token. It's the same as having multiple tokens and cancelling them all at the same time. I've left it as is, because I suspect we'll need to cancel requests more intelligently in the future...

Reply:
I see. I don't mind keep it this way, this is as understandable if not more.